exporter/alibabacloudlogserviceexporter/logsdata_to_logservice.go

```diff
@@ -0,0 +1,205 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package alibabacloudlogserviceexporter
+
+import (
+	"encoding/json"
+	"strconv"
+	"time"
+
+	sls "github.com/aliyun/aliyun-log-go-sdk"
+	"github.com/gogo/protobuf/proto"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.uber.org/zap"
+)
+
+const (
+	slsLogTimeUnixNano   = "timeUnixNano"
+	slsLogSeverityNumber = "severityNumber"
+	slsLogSeverityText   = "severityText"
+	slsLogName           = "name"
+	slsLogContent        = "content"
+	slsLogAttribute      = "attribute"
+	slsLogFlags          = "flags"
+)
+
+func logDataToLogService(logger *zap.Logger, ld pdata.Logs) ([]*sls.Log, int) {
+	numDroppedLogs := 0
+	slsLogs := make([]*sls.Log, 0)
+	rls := ld.ResourceLogs()
+	for i := 0; i < rls.Len(); i++ {
+		rl := rls.At(i)
+		if rl.IsNil() {
+			continue
+		}
+
+		ills := rl.InstrumentationLibraryLogs()
+		for j := 0; j < ills.Len(); j++ {
+			ils := ills.At(j)
+			if ils.IsNil() {
+				continue
+			}
+
+			logs := ils.Logs()
+			for j := 0; j < logs.Len(); j++ {
+				lr := logs.At(j)
+				if lr.IsNil() {
+					continue
+				}
+				slsLog := mapLogRecordToLogService(lr, logger)
+				if slsLog == nil {
+					numDroppedLogs++
+				} else {
+					slsLogs = append(slsLogs, slsLog)
+				}
+			}
+		}
+	}
+
+	return slsLogs, numDroppedLogs
+}
+
+func mapLogRecordToLogService(lr pdata.LogRecord, logger *zap.Logger) *sls.Log {
+	if lr.Body().IsNil() {
+		return nil
+	}
+	var slsLog sls.Log
+
+	// pre alloc, refine if logContent's len > 16
+	slsLog.Contents = make([]*sls.LogContent, 0, 16)
```

Comment:
Nit: It is also possible to preallocate a slice of sls.LogContent and point to them from this slice.

Reply:
done